hist/histv7/inc/ROOT/RAxis.hxx

```diff
@@ -96,6 +96,67 @@ protected:
       return (int)rawbin;
    }
 
+   /// Check if two axis have the same bin borders
+   ///
+   /// Default implementation should work for any RAxis type, but is a little
+   /// bit stupid. RAxis implementations are encouraged to provide optimized
+   /// overrides for common comparison scenarios.
+   virtual bool HasSameBinBordersAs(const RAxisBase& other) const noexcept {
+      // Axis growability (and thus under/overflow bin existence) must match
+      if (CanGrow() != other.CanGrow()) return false;
+
+      // Number of normal bins must match
+      if (GetNBinsNoOver() != other.GetNBinsNoOver()) return false;
+
+      // Left borders of normal bins must match
+      for (int bin: *this) {
+         if (GetBinFrom(bin) != other.GetBinFrom(bin)) return false;
+      }
+
+      // Right border of the last normal bin (aka maximum) must also match
+      if (GetMaximum() != other.GetMaximum()) return false;
+
+      // If all of these checks passed, the two axes have the same bin borders
+      return true;
+   }
+
+   /// If supplementary bin metadata (e.g. bin labels) exists in `this`, make
+   /// sure that `other` provides the same bin metadata.
+   ///
+   /// The implementation can work under the assumption that the number of bins
+   /// and bin borders of `this` and `other` has already been compared equal.
+   ///
+   /// RAxis classes which override a non-default implementation of
+   /// `HasSameBinMetadataAs` should start by calling the parent class'
+   /// implementation of this method, in order to allow "stacking" of bin
+   /// metadata across the class hierarchy.
+   ///
+   /// Adding an override of `HasSameBinMetadataAs` to an RAxis class is
+   /// considered an API-breaking change: subclasses can assume that if their
+   /// parent classes don't override the default implementation of
+   /// `HasSameBinMetadataAs`, they never will.
+   virtual bool HasSameBinMetadataAs(const RAxisBase& /*other*/) const noexcept {
+      return true;
+   }
+
+   /// Semantically equivalent to `HasSameBinMetadataAs(other)`, except that the
+   /// implementation can be optimized under knowledge of the fact that
+   /// `other.HasSameBinMetadataAs(*this)` has previously been called and
+   /// returned a positive result.
+   ///
+   /// RAxis classes which override a non-default implementation of
+   /// `AlsoHasSameBinMetadataAs` should start by calling the parent class'
+   /// implementation of this method, in order to allow "stacking" of bin
+   /// metadata across the class hierarchy.
+   ///
+   /// Adding an override of `AlsoHasSameBinMetadataAs` to an RAxis class is
+   /// considered an API-breaking change: subclasses can assume that if their
+   /// parent classes don't override the default implementation of
+   /// `AlsoHasSameBinMetadataAs`, they never will.
+   virtual bool AlsoHasSameBinMetadataAs(const RAxisBase& other) const noexcept {
+      return HasSameBinMetadataAs(other);
+   }
+
```

Comment:
So, that's admittedly a bit of cursed code. This kind of API would be most cleanly implemented using [multiple dispatch/multi-methods](https://en.wikipedia.org/wiki/Multiple_dispatch), but C++ doesn't have that, and all the classic workarounds that I know of (visitors, hashmap of typeid pairs) rely on assuming knowledge of _every_ child class of `RAxisBase`.

This is something we would like not to do, as being able to extend the RAxis class hierarchy is useful. But under this constraint, I cannot think of a cleaner way to achieve the intended semantics.

Proposals of alternative designs or repriorization of existing design goals (e.g. accepting the alternate stance that it's okay if bin labels don't match) are welcome.

Reply:
Since it seems this comment was not clear enough, I elaborated on the details of the problem in another discussion thread.